File: testdata/fixtures/typescript/CWE-611.ts
// CWE-611: XML External Entity via DOMParser/xml2js
//
// DOMParser and xml2js with defaults resolve external entities. Passing
// user input as the XML string lets an <!ENTITY> become a file read or SSRF.

function handle(req: any) {
  const xml = req.query.xml as string;
  // ruleid: typescript-xxe
  new DOMParser().parseFromString(xml, "text/xml");
  const data = req.body.data as string;
  // ruleid: typescript-xxe
  xml2js.parseString(data);
  const more = req.params.more as string;
  // ruleid: typescript-xxe
  xml2js.parseString(more, {});
}

function safe() {
  // ok: typescript-xxe
  new DOMParser().parseFromString("<root>hello</root>", "text/xml");
  // ok: typescript-xxe
  xml2js.parseString("<root/>");
  // ok: typescript-xxe
  JSON.parse("{}");
}

File: testdata/fixtures/ruby/CWE-611.rb
# CWE-611: XXE via REXML - external entities resolved by default

require 'rexml/document'

def parse_upload
  data = params[:xml]
  # ruleid: ruby-xxe
  REXML::Document.new(data)
end

def parse_body
  xml = request.body.read
  # ruleid: ruby-xxe
  REXML::Document.new(xml)
end

def safe_literal
  # ok: ruby-xxe
  REXML::Document.new("<root>hello</root>")
end

def safe_disabled
  fixed = "<root>safe</root>"
  # ok: ruby-xxe
  REXML::Document.new(fixed)
end
def parse_params
  xml = params[:data]
  # ruleid: ruby-xxe
  REXML::Document.new(xml)
end

def parse_req
  body = request.body.read
  # ruleid: ruby-xxe
  REXML::Document.new(body)
end

def safe3
  # ok: ruby-xxe
  REXML::Document.new("<root>fixed</root>")
end
